design_drafts/async_controller_impl.md

```diff
@@ -0,0 +1,40 @@
+## Implementation of Asynchronous Controllers
+
+
+#### Short description
+
+Asynchronous controllers are an upcoming addition to the ros2_control framework. As the name suggests, they make it possible to run the update calls of these controllers on a separate thread. Asynchronous components are omitted here for the sake of clarity, but they follow a very similar structure.
+Since the intention was to support real-time systems, the standard library's std::atomic interface was used, and due to the inherent difficulty of writing and understanding concurrent lock-free programs, a brief glance at the code might not be enough to figure out if something doesn't work as intended.
+
+For that reason, the purpose of this design draft is to explain how the synchronization works between the Controller Manager's thread and an asynchronous controllers, now to make it easier for others to review the PR of the feature, and later to let the users know if the current implementation is suitable for their use case.
+
+#### Implementation
+
+The controllers running asynchronously are entirely independent of each other, which means that no data sharing occurs between them. Therefore synchronization happens between the Controller Manager's thread and an async controller - so in the following drawings only these two threads will be visualized.
+
+The critical sections are:
+
+* The state interface writes of the Controller Manager's read function
+* The command interface writes of the async controller's update function
+
+The former has to occur and be seen by the async thread before the update, while the latter, ideally, has to finish and be seen before the controller manager's write.
+
+![async_controller](https://user-images.githubusercontent.com/25421074/212979058-5dad88c1-d940-4bc7-ac6f-103711de7210.png)
+
+As shown in the above image, there are four different cases to account for, and these are: 
```

Comment:
Can you also visually show those cases? With small circles in different colors similarly?

```suggestion
The above diagram shows four different concurrency cases to account for:
```

Reply:
Definitely! I'll have to think about how the diagram could be changed, since just adding stuff might further obfuscate it.